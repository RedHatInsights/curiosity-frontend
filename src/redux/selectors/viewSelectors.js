import { createSelectorCreator, defaultMemoize } from 'reselect';
import _get from 'lodash/get';
import _isEqual from 'lodash/isEqual';

/**
 * Create a custom "are objects equal" selector.
 *
 * @private
 * @type {Function}}
 */
const createDeepEqualSelector = createSelectorCreator(defaultMemoize, _isEqual);

/**
 * Return a combined state, props object.
 *
 * @private
 * @param {object} state
 * @param {object} props
 * @param {object} defaultProps
 * @returns {object}
 */
const viewGraphQuery = (state = {}, props, defaultProps = {}) => ({
  graphQuery: {
    ...defaultProps.graphQuery,
    ..._get(state, ['view', 'graphQuery'])
  }
});

/**
 * Create selector, transform combined state, props into a consumable API param/query object.
 *
 * @type {{graphQuery: object}}
 */
const viewSelector = createDeepEqualSelector([viewGraphQuery], viewGraph => ({
  graphQuery: { ...viewGraph.graphQuery }
}));

/**
 * Expose selector instance. For scenarios where a selector is reused across component instances.
 *
 * @param {object} defaultProps
 * @returns {{graphQuery: object}}
 */
const makeViewSelector = defaultProps => (state, props) => ({
  ...viewSelector(state, props, defaultProps)
});

const viewSelectors = {
  view: viewSelector,
  makeView: makeViewSelector
};

export { viewSelectors as default, viewSelectors, viewSelector, makeViewSelector };
