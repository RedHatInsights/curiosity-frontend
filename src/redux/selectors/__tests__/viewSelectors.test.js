import viewSelectors from '../viewSelectors';

describe('ViewSelectors', () => {
  it('should return specific selectors', () => {
    expect(viewSelectors).toMatchSnapshot('selectors');
  });

  it('should pass minimal data on missing a reducer response', () => {
    const state = {};
    expect(viewSelectors.view(state)).toMatchSnapshot('missing reducer error');
  });
});
