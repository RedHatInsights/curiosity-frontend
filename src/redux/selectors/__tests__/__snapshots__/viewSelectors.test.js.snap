// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewSelectors should pass minimal data on missing a reducer response: missing reducer error 1`] = `
Object {
  "query": Object {},
}
`;

exports[`ViewSelectors should return specific selectors: selectors 1`] = `
Object {
  "makeView": [Function],
  "view": [Function],
}
`;
